src/celeritas/global/ActionLauncher.device.hh

```diff
@@ -100,7 +121,38 @@ class ActionLauncher
         auto config = calc_launch_params_(num_threads);
         launch_action_impl<F>
             <<<config.blocks_per_grid, config.threads_per_block, 0, stream>>>(
-                num_threads, call_thread);
+                range(ThreadId{num_threads}), call_thread);
+    }
+
+    //! Launch a Kernel with reduced grid size if tracks are sorted using the
+    //! expected track order strategy.
+    //! TODO: Always use an ActionLauncher instance with the action passed as
+    //! constructor argument
+    void operator()(CoreParams const& params,
+                    CoreState<MemSpace::device> const& state,
+                    ExplicitActionInterface const& action,
+                    F const& call_thread) const
+    {
+        CELER_EXPECT(state.stream_id());
+        if (is_action_sorted(action.order(),
+                             params.init()->host_ref().track_order))
+        {
+            if (Range<ThreadId> threads
+                = state.get_action_range(action.action_id());
+                threads.size())
```

Comment:
Either we should add a `Range::empty()` or write `threads.size() > 0` for clarity